pkg/common/catalog/catalog_test.go

```diff
@@ -8,821 +8,600 @@ import (
+	if tt.mutateRepo != nil {
+		tt.mutateRepo(repo)
+	}
+	if tt.mutatePluginRepo != nil {
+		tt.mutatePluginRepo(pluginRepo)
+	}
+	if tt.mutateServiceRepo != nil {
+		tt.mutateServiceRepo(serviceRepo)
 	}
-	s.builtins = nil
-	s.pluginConfig = nil
-}
 
-func (s *CatalogSuite) AfterTest(suiteName, testName string) {
-	if !s.T().Failed() {
-		return
+	closer, err := catalog.Load(context.Background(), config, repo)
+	if closer != nil {
+		defer closer.Close()
 	}
-	s.T().Logf("Dumping logs for failed test %s/%s:", suiteName, testName)
-	for _, entry := range s.logHook.AllEntries() {
-		s.T().Logf("[%s] %s %v", entry.Level, entry.Message, entry.Data)
+
+	if tt.expectErr != "" {
+		require.EqualError(t, err, tt.expectErr, "load should have failed")
+		assert.Nil(t, closer, "closer should have been nil")
+	} else {
+		require.NoError(t, err, "load should not have failed")
+		assert.NotNil(t, closer, "closer should not have been nil")
+	}
+
+	if tt.expectPluginClient {
+		if assert.NotNil(t, somePlugin, "plugin client should have been initialized") {
+			assert.Equal(t, "test", somePlugin.Name())
+			assert.Equal(t, "SomePlugin", somePlugin.Type())
+			out, err := somePlugin.PluginEcho(context.Background(), "howdy")
+			if assert.NoError(t, err, "call to PluginEcho should have succeeded") {
+				// Assert that the echo response has:
+				// - initial message wrapped by the plugin, then
+				// - wrapped by the name of the plugin as obtained from the host service context, then
+				// - wrapped by the host service
+				assert.Equal(t, "hostService(test(plugin(howdy)))", out)
+			}
+		}
+	} else {
+		assert.Nil(t, somePlugin, "plugin client should not have been initialized")
+	}
+
+	if tt.expectServiceClient {
+		if assert.NotNil(t, someService, "service client should have been initialized") {
+			assert.Equal(t, "test", someService.Name())
+			assert.Equal(t, "SomePlugin", someService.Type())
+			out, err := someService.ServiceEcho(context.Background(), "howdy")
+			if assert.NoError(t, err, "call to ServiceEcho should have succeeded") {
+				// Assert that the echo response has:
+				// - initial message wrapped by the service, then
+				// - wrapped by the name of the plugin as obtained from the host service context, then
+				// - wrapped by the host service
+				assert.Equal(t, "hostService(test(service(howdy)))", out)
+			}
+		}
+	} else {
+		assert.Nil(t, someService, "service client should not have been initialized")
 	}
 }
 
-func (s *CatalogSuite) TestOldPlugin() {
-	require := s.Require()
+func buildTestPlugin(t *testing.T, srcPath string) string {
+	dir := spiretest.TempDir(t)
 
-	path := filepath.Join(s.dir, "oldpluginbin")
-	buildOutput, err := exec.Command("go", "build", "-o", path, "catalog_test_oldplugin.go").CombinedOutput()
+	pluginPath := filepath.Join(dir, "test")
+
+	buildOutput, err := exec.Command("go", "build", "-o", pluginPath, srcPath).CombinedOutput() //nolint: gosec // false positive
```

Comment:
I'm not thrilled with it either, but I also like unit tests to be self contained. Your proposal means that somebody can't just run `go test` in this go package and have the unit tests run. Furthermore, `make test` would need to either build and link the binary every time or detect when rebuilding is necessary, which isn't super straight forward (which is why our Makefile doesn't even attempt to do that).

From my perspective, you are one of three actors:
1) you are working on common catalog
2) you are making a general test pass before pushing to GitHub
3) you are a CI/CD pipeline

For actor 1, you may get a small hit on the first compile until your go package/module cache is warm. This is a one time cost. Anybody doing dev work on common catalog likely already has a warm cache. Subsequent runs are usually a few seconds, depending mostly on your link speed.

For actor 2, you are running `make test` anyway to make sure all unit tests will pass after you push. As mentioned before, this will end up compiling the binary anyway.

For actor 3, you are in the same boat as actor 2.

One thing I can do, with a little bit of effort, is combine the legacy and new plugins into a single binary. That will likely cut the time almost in half, since most of the time is likely spent linking. Let me know if this is an acceptable compromise.

Reply:
I bunched it all into one external plugin. On my machine, with a warm cache, each build takes about 850ms. So this change saves roughly 1.6 seconds.

Reply:
Notice that the test time is roughly equivalent to the existing catalog test code, which also builds an external binary.

Reply:
i.e. 2-3 seconds total (on my machine)

Reply:
Not that I know of. I cleaned my cache locally and ran it against the latest changes and it still completed in about 3.5 seconds, so I'm not exactly sure what yourself and Agustin saw.